dht/sharder.hh

```diff
@@ -28,28 +28,76 @@
 
 namespace dht {
 
+
+// Utilities for sharding ring partition_range:s
+
+// A ring_position range's data is divided into sub-ranges, where each sub-range's data
+// is owned by a single shard. Note that multiple non-overlapping sub-ranges may map to a
+// single shard, and some shards may not receive any sub-range.
+//
+// This module provides utilities for determining the sub-ranges to shard mapping. The utilities
+// generate optimal mappings: each range that you get is the largest possible, so you
+// get the minimum number of ranges possible. You can get many ranges, so operate on them
+// one (or a few) at a time, rather than accumulating them.
+
+// A mapping between a partition_range and a shard. All positions within `ring_range` are
+// owned by `shard`.
+//
+// The classes that return ring_position_range_and_shard make `ring_range` as large as
+// possible (maximizing the number of tokens), so the total number of such ranges is minimized.
+// Successive ranges therefore always have a different `shard` than the previous return.
+// (classes that return ring_position_range_and_shard_and_element can have the same `shard`
+// in successive returns, if `element` is different).
 struct ring_position_range_and_shard {
     dht::partition_range ring_range;
     unsigned shard;
 };
 
+// Incrementally divides a `partition_range` into sub-ranges wholly owned by a single shard.
 class ring_position_range_sharder {
     const sharder& _sharder;
     dht::partition_range _range;
     bool _done = false;
 public:
+    // Initializes the ring_position_range_sharder with a given range to subdivide.
     ring_position_range_sharder(const sharder& sharder, nonwrapping_range<ring_position> rrp)
             : _sharder(sharder), _range(std::move(rrp)) {}
+    // Fetches the next range-shard mapping. When the input range is exhausted, std::nullopt is
+    // returned. The returned ranges are contiguous and non-overlapping, and together span the
+    // entire input range.
     std::optional<ring_position_range_and_shard> next(const schema& s);
 };
 
+// A mapping between a partition_range and a shard (like ring_position_range_and_shard) extended
+// by having a reference to input range index. See ring_position_range_vector_sharder for use.
+//
+// The classes that return ring_position_range_and_shard_and_element make `ring_range` as large as
+// possible (maximizing the number of tokens), so the total number of such ranges is minimized.
+// Successive ranges therefore always have a different `shard` than the previous return.
+// (classes that return ring_position_range_and_shard_and_element can have the same `shard`
+// in successive returns, if `element` is different).
```

Comment:
I think it is not clear enough what "element" is. Is it an index into something? into what? This is not explained. And if it is an index, why not call it "index"?